src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ApiConventionItem.cs

```diff
@@ -0,0 +1,199 @@
+                    var metadataProviders = conventionMethod.GetCustomAttributes(inherit: false)
+                        .OfType<IApiResponseMetadataProvider>()
+                        .ToArray();
+
+                    return new ApiConventionItem(metadataProviders);
+                }
+            }
+
+            return null;
+        }
+
+        private static MethodInfo GetConventionMethod(MethodInfo method, Type conventionType)
+        {
+            foreach (var conventionMethod in conventionType.GetMethods())
+            {
+                if (IsMatch(method, conventionMethod))
+                {
+                    return conventionMethod;
+                }
+            }
+
+            return null;
+        }
+
+        internal static bool IsMatch(MethodInfo methodInfo, MethodInfo conventionMethod)
+        {
+            var methodNameMatchBehavior = GetNameMatchBehavior(conventionMethod);
+            if (!IsNameMatch(methodInfo.Name, conventionMethod.Name, methodNameMatchBehavior))
+            {
+                return false;
+            }
+
+            var methodParameters = methodInfo.GetParameters();
+            var conventionMethodParameters = conventionMethod.GetParameters();
+
+            for (var i = 0; i < conventionMethodParameters.Length; i++)
+            {
+                var conventionParameter = conventionMethodParameters[i];
+                if (conventionParameter.IsDefined(typeof(ParamArrayAttribute)))
+                {
+                    return true;
+                }
+
+                if (methodParameters.Length <= i)
+                {
+                    return false;
+                }
+
+                var nameMatchBehavior = GetNameMatchBehavior(conventionParameter);
+                var typeMatchBehavior = GetTypeMatchBehavior(conventionParameter);
+
+                if (!IsTypeMatch(methodParameters[i].ParameterType, conventionParameter.ParameterType, typeMatchBehavior) ||
+                    !IsNameMatch(methodParameters[i].Name, conventionParameter.Name, nameMatchBehavior))
+                {
+                    return false;
+                }
+            }
+
+            // Ensure convention has at least as many parameters as the method. params convention argument are handled
+            // inside the for loop.
+            return methodParameters.Length == conventionMethodParameters.Length;
+
+            ApiConventionNameMatchBehavior GetNameMatchBehavior(ICustomAttributeProvider attributeProvider)
+            {
+                var attributes = attributeProvider.GetCustomAttributes(inherit: false);
+                for (var i = 0; i < attributes.Length; i++)
+                {
+                    if (attributes[i] is ApiConventionNameMatchAttribute nameMatchAttribute)
+                    {
+                        return nameMatchAttribute.MatchBehavior;
+                    }
+                }
+
+                return ApiConventionNameMatchBehavior.Exact;
```

Comment:
I think we might get some feedback about this - since it's the most restrictive behavior. Consider `SomeAction([TypeMatchBehavior(...)] int param)` this behavior results in the parameter in the action to be named `param`

Reply:
Are you looking for feedback on the set of match behaviors and what they are used for?

You can anticipate what I will say about it. I'll give you a minute to imagine what I will say.

------------

Just because we can doesn't mean we should. What are the set of behaviors we know that we need for the key scenarios?

Reply:
`Prefix` and `Suffix` for names we use. `Any` for parameter names was something we came up as part of the design. For types, `Any`, and `AssignableFrom` would be things I see people using. `Exact` is super unlikely, but it was the easiest to implement.

That said, my statement here was about the default behavior when no attribute is specified. Using Exact sort of seemed ok for now until we get more feedback